### PR TITLE
feat(pandemonium): improve pandemonium

### LINKED ISSUE #57: Improve gateway handling as part of the pre-0.4 refactor
This issue aims to track the updates regarding the pre-0.4-roadmap gateway refactor.

- [x] Introduce a new `HELLO` payload to supply instance data on connection.
  ```js
  {
    "op": "HELLO",
    "d": {
      "heartbeat_interval": Number, // heartbeat interval in milliseconds
      "instance_info": InstanceInfo, // without the ratelimit-info
      "pandemonium_info": PandemoniumConf, // Pandemonium related info including pandemonium ratelimits
    }
  }
  ```
- [x] Improve gateway rate-limiting by not immediately disconnecting clients, but instead sending a warning when you hit the rate-limit. This warning will tell you how long to wait before sending any more gateway payloads. If you send anything in this waiting period you get disconnected.

  A gateway rate-limit warning would look like so:
  ```js
  {
    "op": "RATE_LIMIT",
    "d": {
      "wait": Number // duration to wait in milliseconds
    }
  }
  ```
  
  If you connect to pandemonium while being rate-limited you won't be immediately disconnected and in stead will get a `RATE_LIMIT` payload and it acts as a normal rate-limit (you send a payload while under rate-limit, you get disconnected).

### PR DESCRIPTION
**Pull request by EnokiUN**
## Description

This PR applies the changes proposed in #57 in hopes of improving Pandemonium.

closes #57

## This is a **Code Change**

- [x] Docs have been updated to reflect these changes if necessary.
- [x] Changes have been tested.

### CODE REVIEW

File: todel/src/models/gateway.rs [hunk 1]
```diff
@@ -1,14 +1,127 @@
...
+    ///       "rate_limit": {
+    ///         "reset_after": 10,
+    ///         "limit": 5
+    ///       }
+    ///     }
+    ///   }
+    /// }
+    /// ```
+    Hello {
+        /// The amount of milliseconds your ping interval is supposed to be
```
Comment (ooliver1) [inline todel/src/models/gateway.rs:79]: Wasn't this to be removed?

File: todel/src/models/gateway.rs [hunk 2]
```diff
@@ -1,14 +1,127 @@
...
+        heartbeat_interval: u64,
+        /// The instance's info
+        ///
+        /// This is the same payload you get from the [`get_instance_info`] payload without
+        /// ratelimits
```
Comment (ooliver1) [inline todel/src/models/gateway.rs:84]: https://discord.com/channels/980412957060137001/1048626763615449138/1096240851472101508

Comment (EnokiUN) [reply todel/src/models/gateway.rs:84]: I'm not entirely sure when but I believe we decided to keep it to pandemonium info while working on it in a VC which should provide you with the info you need as to properly interface with the gateway without getting bombarded with god-knows-how-many megabytes of rate-limiting information every time you connect to the gateway

**Review by Chromosomologist — APPROVED** [review]
I believe this is all good. LGTM.

**Pull request merged**